src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1368,52 +1435,378 @@ private sealed class RemainingTestsForCase
+                {
+                    var trueBuilder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
+                    var falseBuilder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
+                    foreach (var other in RemainingTests)
+                    {
+                        other.Filter(builder, test, whenTrueValues, whenFalseValues, out Tests oneTrue, out Tests oneFalse, ref foundExplicitNullTest);
+                        trueBuilder.Add(oneTrue);
+                        falseBuilder.Add(oneFalse);
+                    }
+
+                    whenTrue = Update(trueBuilder);
+                    whenFalse = Update(falseBuilder);
+                }
+                public override Tests RemoveEvaluation(BoundDagEvaluation e)
                 {
-                    result = Hash.Combine((int)d.Kind, result);
+                    var builder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
+                    foreach (var test in RemainingTests)
+                        builder.Add(test.RemoveEvaluation(e));
+
+                    return Update(builder);
                 }
+                public override bool Equals(object? obj)
+                {
+                    if (!(obj is SequenceTests other))
+                        return false;
+
+                    if (this.GetType() != other.GetType())
+                        return false;
+
+                    if (this.RemainingTests.Length != other.RemainingTests.Length)
+                        return false;
 
-                return result;
+                    for (int i = 0, n = this.RemainingTests.Length; i < n; i++)
+                    {
+                        if (!this.RemainingTests[i].Equals(other.RemainingTests[i]))
+                            return false;
+                    }
+
+                    return true;
+                }
+                public override int GetHashCode()
+                {
+                    int length = this.RemainingTests.Length;
+                    int value = Hash.Combine(length, this.GetType().GetHashCode());
+                    for (int i = 0; i < length; i++)
+                        value = Hash.Combine(this.RemainingTests[i].GetHashCode(), value);
+
+                    return value;
+                }
+            }
+
+            /// <summary>
+            /// A sequence of tests that must be performed, each of which must succeed.
+            /// The sequence is deemed to succeed if no element fails.
+            /// </summary>
+            public class AndSequence : SequenceTests
+            {
+                private AndSequence(ImmutableArray<Tests> RemainingTests) : base(RemainingTests) { }
+                public override Tests Update(ArrayBuilder<Tests> remainingTests) => Create(remainingTests);
+                public static Tests Create(ArrayBuilder<Tests> remainingTests)
+                {
+                    for (int i = remainingTests.Count - 1; i >= 0; i--)
```

Comment:
>for (int i = remainingTests.Count - 1; i >= 0; i--) [](start = 20, length = 51)

Why backwards? Same question for `OrSequence`.

Reply:
Because that permits us to remove elements without having to change what the next index we need to handle is.

---
In reply to: [382851602](https://github.com/dotnet/roslyn/pull/41503#discussion_r382851602) [](ancestors = 382851602)